routing_common/transit_graph_data.cpp

```diff
@@ -0,0 +1,586 @@
+  }
+
+  vector<Shape> shapes;
+  for (auto const & shape : m_shapes)
+  {
+    if (shapeIdInEdges.count(shape.GetId()) != 0)
+      shapes.push_back(shape);
+  }
+
+  m_shapes.swap(shapes);
+}
+
+void GraphData::ReadHeader(NonOwningReaderSource & src)
+{
+  FixedSizeDeserializer<NonOwningReaderSource> numberDeserializer(src);
+  numberDeserializer(m_header);
+  CHECK_EQUAL(src.Pos(), m_header.m_stopsOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  CHECK(m_header.IsValid(), ());
+}
+
+void GraphData::ReadStops(NonOwningReaderSource & src)
+{
+  Deserializer<NonOwningReaderSource> deserializer(src);
+  CHECK_EQUAL(src.Pos(), m_header.m_stopsOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  deserializer(m_stops);
+  CHECK_EQUAL(src.Pos(), m_header.m_gatesOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  CHECK(IsValidSortedUnique(m_stops), ());
+}
+
+void GraphData::ReadGates(NonOwningReaderSource & src)
+{
+  Deserializer<NonOwningReaderSource> deserializer(src);
+  CHECK_EQUAL(src.Pos(), m_header.m_gatesOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  deserializer(m_gates);
+  CHECK_EQUAL(src.Pos(), m_header.m_edgesOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  CHECK(IsValidSortedUnique(m_gates), ());
+}
+
+void GraphData::ReadEdges(NonOwningReaderSource & src)
+{
+  Deserializer<NonOwningReaderSource> deserializer(src);
+  CHECK_EQUAL(src.Pos(), m_header.m_edgesOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  deserializer(m_edges);
+  CHECK_EQUAL(src.Pos(), m_header.m_transfersOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  CHECK(IsValidSortedUnique(m_edges), ());
+}
+
+void GraphData::ReadTransfers(NonOwningReaderSource & src)
+{
+  Deserializer<NonOwningReaderSource> deserializer(src);
+  CHECK_EQUAL(src.Pos(), m_header.m_transfersOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  deserializer(m_transfers);
+  CHECK_EQUAL(src.Pos(), m_header.m_linesOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  CHECK(IsValidSortedUnique(m_transfers), ());
+}
+
+void GraphData::ReadLines(NonOwningReaderSource & src)
+{
+  Deserializer<NonOwningReaderSource> deserializer(src);
+  CHECK_EQUAL(src.Pos(), m_header.m_linesOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  deserializer(m_lines);
+  CHECK_EQUAL(src.Pos(), m_header.m_shapesOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  CHECK(IsValidSortedUnique(m_lines), ());
+}
+
+void GraphData::ReadShapes(NonOwningReaderSource & src)
+{
+  Deserializer<NonOwningReaderSource> deserializer(src);
+  CHECK_EQUAL(src.Pos(), m_header.m_shapesOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  deserializer(m_shapes);
+  CHECK_EQUAL(src.Pos(), m_header.m_networksOffset, ("Wrong", TRANSIT_FILE_TAG, "section format."));
+  CHECK(IsValidSortedUnique(m_shapes), ());
+}
+
+void GraphData::ReadNetworks(NonOwningReaderSource & src)
+{
+  Deserializer<NonOwningReaderSource> deserializer(src);
```

Comment:
Again duplication - all ReadX methods (except ReadHeader) create deserializer from src, and applies this deserializer to m_networks. It it possible to create a single method instead, that will take ReaderSource, a field to deserialize and a couple of offsets (where to start and where to end)?

Reply:
Done